lib/concurrent/utility/monotonic_time.rb

```diff
@@ -0,0 +1,147 @@
+module Concurrent
+
+  # Clock that cannot be set and represents monotonic time since
+  # some unspecified starting point.
+  # @!visibility private
+  GLOBAL_MONOTONIC_CLOCK = Class.new {
+
+    if defined?(Process::CLOCK_MONOTONIC)
+      # @!visibility private
+      def get_time
+        Process.clock_gettime(Process::CLOCK_MONOTONIC)
+      end
+    elsif RUBY_PLATFORM == 'java'
+      # @!visibility private
+      def get_time
+        java.lang.System.nanoTime() / 1_000_000_000.0
+      end
+    else
+
+      require 'thread'
+
+      # @!visibility private
+      def initialize
+        @mutex = Mutex.new
+        @correction = 0
+        @last_time = Time.now.to_f
+      end
+
+      # @!visibility private
+      def get_time
+        @mutex.synchronize {
+          @correction ||= 0 # compensating any back time shifts
+          now = Time.now.to_f
+          corrected_now = now + @correction
+          if @last_time < corrected_now
+            @last_time = corrected_now 
+          else
+            @correction += @last_time - corrected_now + 0.000_001
+            @last_time = @correction + now
+          end
+        }
+      end
+    end
+  }.new
```

Comment:
-   minor: there should be `do end` instead of `{}` for multiline blocks
-   Why is the class hidden and anonymous? That seems very unusual. 
-   Why usual pattern for different implementations is not followed? (`JavaMonotonicClock` etc.)